docs/design/hooks.md

```diff
@@ -0,0 +1,219 @@
+
+### Proposed Implementation
+NFS-Provisioner will use user-provided Configmap to learn hooks configuration. This hooks will be executed on volume provisioning or deleting events to set the given information on provided resources. 
+
+### High Level Design
+User will deploy nfs-provisioner with Configmap having information about hook information. This Configmap should exists in the same namespace in which nfs-provisioner is deployed. Sample nfs-provisioner deployment config is as below:
+
+```yaml
+    spec:
+      serviceAccountName: openebs-maya-operator
+      containers:
+      - name: openebs-provisioner-nfs
+        image: openebs/provisioner-nfs:ci
+        env:
+        - name: OPENEBS_IO_HOOK_CONFIG
+          value: "nfs-hook"
+        livenessProbe:
+          exec:
+            command:
+            - sh
+            - -c
+            - test `pgrep "^provisioner-nfs.*"` = 1
+          initialDelaySeconds: 30
+          periodSeconds: 60
+```
+
+User need to provide Configmap name as value of `OPENEBS_IO_HOOK_CONFIG` environment variable.
+
+### Low Level Design
+#### Hook Definition
+NFS Provisioner uses `Provisioner` to provision NFS PV. Existing `Provisioner` definition needs to be extended to use following hooks definition.
+
+```go
+// ResourceType defines type of resource
+type ResourceType string
+
+const (
+	// Type of resources created by nfs-provisioner
+	ResourceBackendPVC          ResourceType = "BackendPVC"
+	ResourceBackendPV           ResourceType = "BackendPV"
+	ResourceNFSService          ResourceType = "NfsService"
+	ResourceNFSPV               ResourceType = "NfsPV"
+	ResourceNFSServerDeployment ResourceType = "NfsServerDeployment"
+
+	// ResourceAll represent the all above resources
+	ResourceAll ResourceType = "All"
+)
+
+// HookActionType defines type of action for annotation and finalizer
+type HookActionType string
+
+const (
+	HookActionAdd    HookActionType = "Add"
+	HookActionRemove HookActionType = "Remove"
+)
+
+// ProvisionerEventType defines type of events on which hook needs to be executed
+type ProvisionerEventType string
+
+const (
+	ProvisionerEventCreate ProvisionerEventType = "Create"
+	ProvisionerEventDelete ProvisionerEventType = "Delete"
+)
+
+// HookConfig represent the hook to be executed by nfs-provisioner
+type HookConfig struct {
+	// Name represent hook name
+	Name string `json:"name"`
+
+	// List of Annotations, mapped to type of resources, needs to be added on given resource
+	Annotations map[ResourceType][]string `json:"annotations"`
+
+	// List of Finalizers, mapped to type of resources, needs to be added on given resource
+	Finalizers map[ResourceType][]string `json:"finalizers"`
+
+	// Event defines provisioning event on which
+	// given hook action needs to be executed
+	Event ProvisionerEventType `json:"provisioningEvent"`
+
+	// Action represent the type of hook action, i.e HookActionAdd or HookActionRemove
+	Action HookActionType `json:"hookAction"`
+}
+
```

Comment:
Can we have high-level Config? In future, this config will help to add tolerations, imagePullSecrets e.t.c
```go
type Config struct {
      Tolerations TolerationsStruct
      Hooks          []HookConfig
      ImagePullSecrets ImagePullSecrets
}
```

Reply:
Toleration/ImagePullSecrets can be added in a similar way to Annotation. Since hook will be executed on a specific event and on the list of objects, we can target Toleration/PullSecrets on required objects.